### PR TITLE
Multiple blade element model

### PR DESCRIPTION
**Pull request by courtin**
Adds propeller blade element model

### CODE REVIEW

**Comment by mjburton11** [comment]
Is this the multi-element model?

**Comment by courtin** [comment]
yes.  The name is very misleading....sorry

File: gpkitmodels/GP/aircraft/prop/propeller.py [hunk 1]
```diff
@@ -55,6 +55,123 @@ def setup(self, static, state):
...
+                        blade.omega == omega,
+                        blade["r"][0] == static.R/(2.*N)]
+
+        
+        with SignomialsEnabled():
+            for n in range(1,N):
+                constraints += [TCS([blade["r"][n] >= blade["r"][n-1] + static.R/N]),
+                                blade["eta_i"][n] == blade["eta_i"][n-1],
+                                ]
+            constraints += [TCS([T <= sum(blade["dT"])]),
```
Comment (bqpd) [inline gpkitmodels/GP/aircraft/prop/propeller.py]: use dot notation instead of square brackets?

File: gpkitmodels/GP/aircraft/prop/propeller.py [hunk 2]
```diff
@@ -55,6 +55,123 @@ def setup(self, static, state):
...
+        with Vectorize(N):
+            blade = Blade_Element_Performance(static, state)
+
+
+        constraints = [blade.dr == static.R/(N),
+                        blade.omega == omega,
+                        blade["r"][0] == static.R/(2.*N)]
+
+        
+        with SignomialsEnabled():
```
Comment (bqpd) [inline gpkitmodels/GP/aircraft/prop/propeller.py]: is this an SP? It's currently in the GP folder of gplibrary

**Comment by courtin** [comment]
test this please

**Comment by courtin** [comment]
test this please

**Comment by courtin** [comment]
@mjburton11 @bqpd  I'm not sure why this is failing tests; it looks like the file it's failing to find (dae51_fitdata.csv) is being tracked in the correct location on this branch.  Does it need to be added somewhere to install correctly?

File: gpkitmodels/SP/aircraft/prop/propeller.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    eta_i                   [-]         local induced efficiency
+    dQ                      [N*m]       torque
+    omega                   [rpm]       propeller rotation rate 
+    Wa                      [m/s]       Axial total relative velocity
+    Wt                      [m/s]       Tangential total relative velocity
+    Wr                      [m/s]       Total relative velocity
+    va                      [m/s]       Axial induced velocity
+    vt                      [m/s]       Tangential induced velocity
+    G                       [m^2/s]     Circulation
+    cl                     [-]          local lift coefficient
```
Comment (mjburton11) [inline gpkitmodels/SP/aircraft/prop/propeller.py]: fix alignments

File: gpkitmodels/SP/aircraft/prop/propeller.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+from numpy import pi
+from gpkit import Model, Variable,Vectorize,parse_variables, SignomialsEnabled, SignomialEquality
+from gpkit.constraints.tight import Tight as TCS
+from gpfit.fit_constraintset import XfoilFit
+import os
+import pandas as pd
+
+
+
+class Blade_Element_Performance(Model):
```
Comment (mjburton11) [inline gpkitmodels/SP/aircraft/prop/propeller.py]: Can we rename this to BladeElementPerf

File: gpkitmodels/SP/aircraft/prop/propeller.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+        R       = static.R
+        mu      = state.mu
+        path = os.path.dirname(__file__)
+        fd = pd.read_csv(path + os.sep + "dae51_fitdata.csv").to_dict(
+            orient="records")[0]
+        c = static.c
+        constraints = [TCS([Wa>=V + va]),
+                        TCS([Wt + vt<=omega*r]),
+                        TCS([G == (1./2.)*Wr*c*cl]),
+                        F == (2./pi)*(1.01116*f**.0379556)**(1./.1), #This is the GP fit of arccos(exp(-f))
```
Comment (mjburton11) [inline gpkitmodels/SP/aircraft/prop/propeller.py]: Is there a reason you didn't make this a fit constraint set?

Comment (mjburton11) [reply gpkitmodels/SP/aircraft/prop/propeller.py]: also why is it raised to 1./1.

File: gpkitmodels/SP/aircraft/prop/propeller.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+                        
+                        XfoilFit(fd, cd, [cl,Re], name="polar"),
+                        
+                        cl <= cl_max
+                    ]
+        with SignomialsEnabled():
+            constraints += [SignomialEquality(Wr**2,(Wa**2+Wt**2)),
+                            
+                            TCS([dT <= rho*B*G*(Wt-eps*Wa)*dr]),
+                            #SignomialEquality(vt**2*F**2*(1.+(4.*lam_w*R/(pi*B*r))**2),(B*G/(4.*pi*r))**2),
```
Comment (mjburton11) [inline gpkitmodels/SP/aircraft/prop/propeller.py]: can we delete commented out lines?

File: gpkitmodels/SP/aircraft/prop/propeller.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+                        eta_i == (V/(omega*r))*(Wt/Wa),
+                        TCS([f+(r/R)*B/(2*lam_w) <= (B/2.)*(1./lam_w)]),
+                        
+                        XfoilFit(fd, cd, [cl,Re], name="polar"),
+                        
+                        cl <= cl_max
+                    ]
+        with SignomialsEnabled():
+            constraints += [SignomialEquality(Wr**2,(Wa**2+Wt**2)),
+
```
Comment (mjburton11) [inline gpkitmodels/SP/aircraft/prop/propeller.py]: delete blank spaces

File: gpkitmodels/SP/aircraft/prop/propeller.py [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+                            TCS([vt**2*F**2*(1.+(4.*lam_w*R/(pi*B*r))**2)>=(B*G/(4.*pi*r))**2]),
+                            #SignomialEquality(f+(r/R)*B/(2*lam_w),(B/2.)*(1./lam_w))
+                            
+                    
+            ]
+        return constraints, state
+
+
+
+class MultiElementProp(Model):
```
Comment (mjburton11) [inline gpkitmodels/SP/aircraft/prop/propeller.py]: I think this should be PropPerf or QPropPerf.  Multielement is maybe a little ambiguous

File: gpkitmodels/SP/aircraft/prop/propeller.py [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+        with SignomialsEnabled():
+            for n in range(1,N):
+                constraints += [TCS([blade.r[n] >= blade.r[n-1] + static.R/N]),
+                                blade.eta_i[n] == blade.eta_i[n-1],
+                                ]
+            constraints += [TCS([T <= sum(blade.dT)]),
+                            TCS([Q >= sum(blade.dQ)]),
+                            eta == state.V*T/(omega*Q),
+                            blade.M[-1] <= Mtip,
+                            static.T_m>=T,
```
Comment (mjburton11) [inline gpkitmodels/SP/aircraft/prop/propeller.py]: spaces before and after operators

File: gpkitmodels/SP/aircraft/prop/propeller.py [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+        with Vectorize(N):
+            blade = Blade_Element_Performance(static, state)
+
+
+        constraints = [blade.dr == static.R/(N),
+                        blade.omega == omega,
+                        blade.r[0] == static.R/(2.*N)]
+
+        
+        with SignomialsEnabled():
```
Comment (mjburton11) [inline gpkitmodels/SP/aircraft/prop/propeller.py]: It seems like there's GP constraints in here. can we move them above?

File: gpkitmodels/SP/aircraft/prop/propeller.py (hunk 5 above)
Comment (courtin) [reply gpkitmodels/SP/aircraft/prop/propeller.py]: simplicity of the fit; it's raised to 1/.1 (otherwise known as 10)

**Comment by mjburton11** [comment]
@courtin add csv file to setup so that should fix the problem

**Comment by courtin** [comment]
@mjburton11 should be fixed now

**Comment by courtin** [comment]
test models please

**Comment by mjburton11** [comment]
@courtin merge when ready!

**Pull request merged**